README.md
# CIS Docker Benchmark - InSpec Profile

[![Build Status](http://img.shields.io/travis/dev-sec/cis-docker-benchmark.svg)][1]
[![Supermarket](https://img.shields.io/badge/InSpec%20Profile-CIS%20Docker%20Benchmark-brightgreen.svg)](https://supermarket.chef.io/tools/cis-docker-benchmark)
[![Gitter Chat](https://badges.gitter.im/Join%20Chat.svg)][2]

## Description

This [InSpec](https://github.com/chef/inspec) compliance profile implement the [CIS Docker 1.13.0 Benchmark](https://downloads.cisecurity.org/) in an automated way to provide security best-practice tests around Docker daemon and containers in a production environment.

InSpec is an open-source run-time framework and rule language used to specify compliance, security, and policy requirements for testing any node in your infrastructure.

## Requirements

* at least [InSpec](http://inspec.io/) version 2.3.23
* Docker 1.13+

### Platform

* Debian 8
* Ubuntu 16.04
* CentOS 7

## Attributes

We use a yml attribute file to steer the configuration, the following options are available:

* `trusted_user: vagrant`
  define trusted user to control Docker daemon.
* `authorization_plugin: authz-broker`
  define authorization plugin to manage access to Docker daemon.
* `log_driver: syslog`
  define preferable way to store logs.
* `log_opts: /syslog-address/`
  define Docker daemon log-opts.
* `registry_cert_path: /etc/docker/certs.d`
  directory contains various Docker registry directories.
* `registry_name: /etc/docker/certs.d/registry_hostname:port`
  directory contain certificate certain Docker registry.
* `registry_ca_file: /etc/docker/certs.d/registry_hostname:port/ca.crt`
  certificate file for a certain Docker registry certificate files.
* `container_user: vagrant`
  define user within containers.
* `app_armor_profile: docker-default`
  define apparmor profile for Docker containers.
* `selinux_profile: /label\:level\:s0-s0\:c1023/`
  define SELinux profile for Docker containers.
* `container_capadd: null`
  define needed capabilities for containers. example: `container_capadd: NET_ADMIN,SYS_ADMIN`
* `managable_container_number: 25`
  keep number of containers on a host to a manageable total.
* `daemon_tlscacert : /etc/docker/ssl/ca.pem`
  configure the certificate authority.
* `daemon_tlscert: /etc/docker/ssl/server_cert.pem`
  configure the server certificate.
* `daemon_tlskey: /etc/docker/ssl/server_key.pem`
  configure the server key.
* `swarm_mode: inactive`
  configure the swarm mode.
* `swarm_max_manager_nodes: 3`
  configure the maximum number of swarm leaders.
* `swarm_port: 2377`
  configure the swarm port.
* `benchmark_version`
  to execute also the old controls from previous benchmarks, e.g. set it to 1.12.0 to execute also the tests from cis-benchmark-1.12.0 (which is the default).

These settings can be overridden using an attributes file (e.g. --attrs <attributefile.yml>). See [sample_attributes.yml](sample_attributes.yml) as an example.

## Usage

InSpec makes it easy to run your tests wherever you need. More options listed here: [InSpec cli](http://inspec.io/docs/reference/cli/)

```sh
# run profile locally
$ git clone https://github.com/dev-sec/cis-docker-benchmark
$ inspec exec cis-docker-benchmark

# run profile locally and directly from Github
$ inspec exec https://github.com/dev-sec/cis-docker-benchmark

# run profile on remote host via SSH
inspec exec cis-docker-benchmark -t ssh://user@hostname -i /path/to/key

# run profile on remote host via SSH with sudo
inspec exec cis-docker-benchmark -t ssh://user@hostname -i /path/to/key --sudo

# run profile on remote host via SSH with sudo and define attribute value
inspec exec cis-docker-benchmark --attrs sample_attributes.yml

# run profile direct from inspec supermarket
inspec supermarket exec dev-sec/cis-docker-benchmark -t ssh://user@hostname --key-files private_key --sudo
```

### Run individual controls

In order to verify individual controls, just provide the control ids to InSpec:

```sh
inspec exec cis-docker-benchmark --controls 'cis-docker-benchmark-1.4 cis-docker-benchmark-1.5'
```

## Contributors + Kudos

* Patrick Muench [atomic111](https://github.com/atomic111)
* Dominik Richter [arlimus](https://github.com/arlimus)
* Christoph Hartmann [chris-rock](https://github.com/chris-rock)

## License and Author

* Author:: Patrick Muench <patrick.muench1111@gmail.com>
* Author:: Christoph Hartmann <chris@lollyrock.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

<http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[1]: http://travis-ci.org/dev-sec/cis-docker-benchmark
[2]: https://gitter.im/dev-sec/general
[3]: https://downloads.cisecurity.org/


CHANGELOG.md
# Changelog

## [2.1.4](https://github.com/dev-sec/cis-docker-benchmark/tree/2.1.4) (2023-05-02)

[Full Changelog](https://github.com/dev-sec/cis-docker-benchmark/compare/2.1.3...2.1.4)

**Implemented enhancements:**

- use centralised issue templates and workflows [\#77](https://github.com/dev-sec/cis-docker-benchmark/pull/77) ([schurzi](https://github.com/schurzi))

**Fixed bugs:**

- Wrong placement of flag in command [\#72](https://github.com/dev-sec/cis-docker-benchmark/issues/72)
- Wrong placement of flag in command  [\#73](https://github.com/dev-sec/cis-docker-benchmark/pull/73) ([nnickie23](https://github.com/nnickie23))

**Merged pull requests:**

- add spellchecking with codespell [\#79](https://github.com/dev-sec/cis-docker-benchmark/pull/79) ([schurzi](https://github.com/schurzi))
- Configure Renovate [\#78](https://github.com/dev-sec/cis-docker-benchmark/pull/78) ([renovate[bot]](https://github.com/apps/renovate))
- Change linting to Cookstyle [\#75](https://github.com/dev-sec/cis-docker-benchmark/pull/75) ([schurzi](https://github.com/schurzi))

## [2.1.3](https://github.com/dev-sec/cis-docker-benchmark/tree/2.1.3) (2022-01-12)

[Full Changelog](https://github.com/dev-sec/cis-docker-benchmark/compare/2.1.2...2.1.3)

**Merged pull requests:**

- use input instead of attribute [\#74](https://github.com/dev-sec/cis-docker-benchmark/pull/74) ([micheelengronne](https://github.com/micheelengronne))
- update release workflow to match other projects [\#71](https://github.com/dev-sec/cis-docker-benchmark/pull/71) ([schurzi](https://github.com/schurzi))
- Missing words "certificate authority." [\#70](https://github.com/dev-sec/cis-docker-benchmark/pull/70) ([adamoutler](https://github.com/adamoutler))

## [2.1.2](https://github.com/dev-sec/cis-docker-benchmark/tree/2.1.2) (2020-06-18)

[Full Changelog](https://github.com/dev-sec/cis-docker-benchmark/compare/2.1.1...2.1.2)

**Merged pull requests:**

- version alignment [\#68](https://github.com/dev-sec/cis-docker-benchmark/pull/68) ([micheelengronne](https://github.com/micheelengronne))

## [2.1.1](https://github.com/dev-sec/cis-docker-benchmark/tree/2.1.1) (2020-06-18)

[Full Changelog](https://github.com/dev-sec/cis-docker-benchmark/compare/2.1.0...2.1.1)

**Closed issues:**

- Examples not working [\#58](https://github.com/dev-sec/cis-docker-benchmark/issues/58)
- incompatible character encodings: UTF-8 and ASCII-8BIT [\#51](https://github.com/dev-sec/cis-docker-benchmark/issues/51)

**Merged pull requests:**

- github release action [\#67](https://github.com/dev-sec/cis-docker-benchmark/pull/67) ([micheelengronne](https://github.com/micheelengronne))
- Update Inspec.yml [\#66](https://github.com/dev-sec/cis-docker-benchmark/pull/66) ([MoisesTapia](https://github.com/MoisesTapia))
- Removed trailing slashes in 1.8, 1.9 [\#63](https://github.com/dev-sec/cis-docker-benchmark/pull/63) ([presidenten](https://github.com/presidenten))
- Remove .gitkeep file [\#62](https://github.com/dev-sec/cis-docker-benchmark/pull/62) ([james-stocks](https://github.com/james-stocks))
- Simple fix for \#58 [\#61](https://github.com/dev-sec/cis-docker-benchmark/pull/61) ([commjoen](https://github.com/commjoen))
- Update issue templates [\#57](https://github.com/dev-sec/cis-docker-benchmark/pull/57) ([rndmh3ro](https://github.com/rndmh3ro))
- unified attributes [\#56](https://github.com/dev-sec/cis-docker-benchmark/pull/56) ([chris-rock](https://github.com/chris-rock))
- Removed unneeded processing step [\#55](https://github.com/dev-sec/cis-docker-benchmark/pull/55) ([tstuber](https://github.com/tstuber))

## [2.1.0](https://github.com/dev-sec/cis-docker-benchmark/tree/2.1.0) (2018-04-20)

[Full Changelog](https://github.com/dev-sec/cis-docker-benchmark/compare/2.0.0...2.1.0)

**Closed issues:**

- method\_missing: undefined local variable or method docker [\#50](https://github.com/dev-sec/cis-docker-benchmark/issues/50)
- uploading cis docker profile to chef compliance [\#46](https://github.com/dev-sec/cis-docker-benchmark/issues/46)

**Merged pull requests:**

- 2.1.0 [\#54](https://github.com/dev-sec/cis-docker-benchmark/pull/54) ([chris-rock](https://github.com/chris-rock))
- Fix utf8 truncated output [\#53](https://github.com/dev-sec/cis-docker-benchmark/pull/53) ([aschmidt75](https://github.com/aschmidt75))
- update inspec version to 2.0 [\#52](https://github.com/dev-sec/cis-docker-benchmark/pull/52) ([atomic111](https://github.com/atomic111))
- Fixes \#37 prevent NoMethodError when no hosts available [\#49](https://github.com/dev-sec/cis-docker-benchmark/pull/49) ([Nowheresly](https://github.com/Nowheresly))
- name correct minimum inspec version [\#47](https://github.com/dev-sec/cis-docker-benchmark/pull/47) ([chris-rock](https://github.com/chris-rock))
- update changelog [\#45](https://github.com/dev-sec/cis-docker-benchmark/pull/45) ([chris-rock](https://github.com/chris-rock))

## [2.0.0](https://github.com/dev-sec/cis-docker-benchmark/tree/2.0.0) (2017-11-24)

[Full Changelog](https://github.com/dev-sec/cis-docker-benchmark/compare/1.3.1...2.0.0)

**Closed issues:**

- Verify enable content trust per-shell or per-invocation check [\#44](https://github.com/dev-sec/cis-docker-benchmark/issues/44)
- load_with_context': undefined method `each' for nil:NilClass \(NoMethodError\) exception in cis-docker-benchmark-master/controls/container\_runtime.rb:194 [\#37](https://github.com/dev-sec/cis-docker-benchmark/issues/37)
- use own control number scheme [\#25](https://github.com/dev-sec/cis-docker-benchmark/issues/25)
- Update to CIS 1.13 [\#24](https://github.com/dev-sec/cis-docker-benchmark/issues/24)

**Merged pull requests:**

- Update to CIS Docker Benchmark 1.13.0 [\#43](https://github.com/dev-sec/cis-docker-benchmark/pull/43) ([atomic111](https://github.com/atomic111))
- correct the maintainer and email in inspec.yml [\#42](https://github.com/dev-sec/cis-docker-benchmark/pull/42) ([atomic111](https://github.com/atomic111))

controls/docker_host_os_level1.rb
# encoding: utf-8
# frozen_string_literal: true
#
# Copyright 2016, Patrick Muench
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# author: Christoph Hartmann
# author: Dominik Richter
# author: Patrick Muench

title 'CIS Docker Benchmark - Level 1 - Linux Host OS'

TRUSTED_USER = attribute(
  'trusted_user',
  description: 'define trusted user to control Docker daemon. cis-docker-benchmark-1.6',
  default: 'vagrant'
)

MANAGEABLE_CONTAINER_NUMBER = attribute(
  'managable_container_number',
  description: 'keep number of containers on a host to a manageable total. cis-docker-benchmark-6.5',
  default: 25
)

# check if docker exists
only_if do
  command('docker').exist?
end

control 'cis-docker-benchmark-1.1' do
  impact 1.0
  title 'Create a separate partition for containers'
  desc 'All Docker containers and their data and metadata is stored under /var/lib/docker directory. By default, /var/lib/docker would be mounted under / or /var partitions based on availability.'

  tag 'host'
  tag cis: 'docker:1.1'
  tag level: 1
  ref 'Docker storage recommendation', url: 'http://www.projectatomic.io/docs/docker-storage-recommendation/'

  describe mount('/var/lib/docker') do
    it { should be_mounted }
  end
end

control 'cis-docker-benchmark-1.2' do
  impact 1.0
  title 'Use the updated Linux Kernel'
  desc 'Docker in daemon mode has specific kernel requirements. A 3.10 Linux kernel is the minimum requirement for Docker.'

  tag 'host'
  tag cis: 'docker:1.2'
  tag level: 1
  ref 'Check kernel dependencies', url: 'https://docs.docker.com/engine/installation/binaries/#check-kernel-dependencies'
  ref 'Installation list', url: 'https://docs.docker.com/engine/installation/#installation-list'

  kernel_version = command('uname -r | grep -o \'^\w\.\w*\.\w*\'').stdout
  kernel_compare = Gem::Version.new('3.10') <= Gem::Version.new(kernel_version)

  describe kernel_compare do
    it { should eq true }
  end
end

control 'cis-docker-benchmark-1.3' do
  impact 1.0
  title 'Harden the container host'
  desc 'Containers run on a Linux host. A container host can run one or more containers. It is of utmost importance to harden the host to mitigate host security misconfiguration'

  tag 'host'
  tag cis: 'docker:1.3'
  tag level: 1
  ref 'Hardening Framework dev-sec.io', url: 'http://dev-sec.io'
  ref 'Docker security article', url: 'https://docs.docker.com/engine/security/security/'
  ref 'CIS Benchmarks', url: 'https://benchmarks.cisecurity.org/downloads/multiform/index.cfm'
  ref 'grsecurity', url: 'https://grsecurity.net/'
  ref 'grsecurity Wiki', url: 'https://en.wikibooks.org/wiki/Grsecurity'
  ref 'Homepage of The PaX Team', url: 'https://pax.grsecurity.net/'
  ref 'PAX Wiki', url: 'http://en.wikipedia.org/wiki/PaX'
end

control 'cis-docker-benchmark-1.4' do
  impact 1.0
  title 'Remove all non-essential services from the host'
  desc 'Ensure that the host running the docker daemon is running only the essential services.'

  tag 'host'
  tag cis: 'docker:1.4'
  tag level: 1
  ref 'Containers & Docker: How Secure Are They?', url: 'https://blog.docker.com/2013/08/containers-docker-how-secure-are-they/'
end

control 'cis-docker-benchmark-1.5' do
  impact 1.0
  title 'Keep Docker up to date'
  desc 'The docker container solution is evolving to maturity and stability at a rapid pace. Like any other software, the vendor releases regular updates for Docker software that address security vulnerabilities, product bugs and bring in new functionality.'

  tag 'host'
  tag cis: 'docker:1.5'
  tag level: 1
  ref 'Docker installation', url: 'https://docs.docker.com/installation/'
  ref 'Docker releases', url: 'https://github.com/docker/docker/releases/latest'

  docker_server_version = command('docker version --format \'{{.Server.Version}}\'').stdout
  docker_server_compare = Gem::Version.new('17.03') <= Gem::Version.new(docker_server_version)

  docker_client_version = command('docker version --format \'{{.Client.Version}}\'').stdout
  docker_client_compare = Gem::Version.new('17.03') <= Gem::Version.new(docker_client_version)

  describe docker_server_compare do
    it { should eq true }
  end

  describe docker_client_compare do
    it { should eq true }
  end
end

control 'cis-docker-benchmark-1.6' do
  impact 1.0
  title 'Only allow trusted users to control Docker daemon'
  desc 'The Docker daemon currently requires \'root\' privileges. A user added to the \'docker\' group gives him full \'root\' access rights'

  tag 'host'
  tag cis: 'docker:1.6'
  tag level: 1
  ref 'On Docker security: docker group considered harmful', url: 'https://www.andreas-jung.com/contents/on-docker-security-docker-group-considered-harmful'
  ref 'Why we do not let non-root users run Docker in CentOS, Fedora, or RHEL', url: 'http://www.projectatomic.io/blog/2015/08/why-we-dont-let-non-root-users-run-docker-in-centos-fedora-or-rhel/'

  describe group('docker') do
    it { should exist }
  end

  describe etc_group.where(group_name: 'docker') do
    its('users') { should include TRUSTED_USER }
  end
end

control 'cis-docker-benchmark-benchmark-1.7' do
  impact 1.0
  title 'Audit docker daemon'
  desc 'Apart from auditing your regular Linux file system and system calls, audit Docker daemon as well. Docker daemon runs with \'root\' privileges. It is thus necessary to audit its activities and usage.'

  tag 'host'
  tag cis: 'docker:1.7'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'

  describe auditd_rules do
    its(:lines) { should include('-w /usr/bin/docker -p rwxa -k docker') }
  end
  describe service('auditd') do
    it { should be_installed }
    it { should be_enabled }
    it { should be_running }
  end
end

control 'cis-docker-benchmark-1.8' do
  impact 1.0
  title 'Audit Docker files and directories - /var/lib/docker'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. /var/lib/docker is one such directory. It holds all the information about containers. It must be audited.'

  tag 'host'
  tag cis: 'docker:1.8'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'

  describe auditd_rules do
    its(:lines) { should include('-w /var/lib/docker/ -p rwxa -k docker') }
  end
end

control 'cis-docker-benchmark-1.9' do
  impact 1.0
  title 'Audit Docker files and directories - /etc/docker'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. /etc/docker is one such directory. It holds various certificates and keys used for TLS communication between Docker daemon and Docker client. It must be audited.'

  tag 'host'
  tag cis: 'docker:1.9'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'

  describe auditd_rules do
    its(:lines) { should include('-w /etc/docker/ -p rwxa -k docker') }
  end
end

control 'cis-docker-benchmark-1.10' do
  impact 1.0
  title 'Audit Docker files and directories - docker.service'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. docker.service is one such file. The docker.service file might be present if the daemon parameters have been changed by an administrator. It holds various parameters for Docker daemon. It must be audited, if applicable.'

  tag 'host'
  tag cis: 'docker:1.10'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'

  if docker.path
    rule = '-w ' + docker.path + ' -p rwxa -k docker'
    describe auditd_rules do
      its(:lines) { should include(rule) }
    end
  else
    describe 'audit docker service' do
      skip 'Cannot determine docker path'
    end
  end
end

control 'cis-docker-benchmark-1.11' do
  impact 1.0
  title 'Audit Docker files and directories - docker.socket'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. docker.socket is one such file. It holds various parameters for Docker daemon socket. It must be audited, if applicable.'

  tag 'host'
  tag cis: 'docker:1.11'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'

  if docker.socket
    rule = '-w ' + docker.socket + ' -p rwxa -k docker'
    describe auditd_rules do
      its(:lines) { should include(rule) }
    end
  else
    describe 'audit docker service' do
      skip 'Cannot determine docker socket'
    end
  end
end

control 'cis-docker-benchmark-1.12' do
  impact 1.0
  title 'Audit Docker files and directories - /etc/default/docker'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. /etc/default/docker is one such file. It holds various parameters for Docker daemon. It must be audited, if applicable.'

  tag 'host'
  tag cis: 'docker:1.12'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'

  only_if { os[:family] != 'centos' }
  describe auditd_rules do
    its(:lines) { should include('-w /etc/default/docker -p rwxa -k docker') }
  end
end

control 'cis-docker-benchmark-1.13' do
  impact 1.0
  title 'Audit Docker files and directories - /etc/docker/daemon.json'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. /etc/docker/daemon.json is one such file. It holds various parameters for Docker daemon. It must be audited, if applicable.'

  tag 'host'
  tag cis: 'docker:1.13'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'
  ref 'Daemon configuration', url: 'https://docs.docker.com/engine/reference/commandline/daemon/#daemon-configuration-file'

  describe auditd_rules do
    its(:lines) { should include('-w /etc/docker/daemon.json -p rwxa -k docker') }
  end
end

control 'cis-docker-benchmark-1.14' do
  impact 1.0
  title 'Audit Docker files and directories - /usr/bin/docker-containerd'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. /usr/bin/docker-containerd is one such file. Docker now relies on containerd and runC to spawn containers. It must be audited, if applicable.'

  tag 'host'
  tag cis: 'docker:1.14'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'
  ref 'Containerd integration', url: 'https://github.com/docker/docker/pull/20662'
  ref 'Containerd tools', url: 'https://containerd.tools/'

  describe auditd_rules do
    its(:lines) { should include('-w /usr/bin/docker-containerd -p rwxa -k docker') }
  end
end

control 'cis-docker-benchmark-1.15' do
  impact 1.0
  title 'Audit Docker files and directories - /usr/bin/docker-runc'
  desc 'Apart from auditing your regular Linux file system and system calls, audit all Docker related files and directories. Docker daemon runs with \'root\' privileges. Its behavior depends on some key files and directories. /usr/bin/docker-runc is one such file. Docker now relies on containerd and runC to spawn containers. It must be audited, if applicable.'

  tag 'host'
  tag cis: 'docker:1.15'
  tag level: 1
  ref 'System auditing', url: 'https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Security_Guide/chap-system_auditing.html'
  ref 'Containerd integration', url: 'https://github.com/docker/docker/pull/20662'
  ref 'Containerd tools', url: 'https://containerd.tools/'
  ref 'Opencontainers runc repository', url: 'https://github.com/opencontainers/runc'

  describe auditd_rules do
    its(:lines) { should include('-w /usr/bin/docker-runc -p rwxa -k docker') }
  end
end

control 'cis-docker-benchmark-6.4' do
  impact 1.0
  title 'Avoid image sprawl'
  desc 'Do not keep a large number of container images on the same host. Use only tagged images as appropriate.'

  tag 'host'
  tag cis: 'docker:6.4'
  tag level: 1
  ref 'http://craiccomputing.blogspot.de/2014/09/clean-up-unused-docker-containers-and.html'
  ref 'https://forums.docker.com/t/command-to-remove-all-unused-images/20/7'
  ref 'https://github.com/docker/docker/issues/9054'
  ref 'https://docs.docker.com/engine/reference/commandline/cli/#rmi'
  ref 'https://docs.docker.com/engine/reference/commandline/cli/#pull'
  ref 'https://github.com/docker/docker/pull/11109'

  instantiated_images = command('docker ps -qa | xargs docker inspect -f \'{{.Image}}\'').stdout.split
  all_images = command('docker images -q --no-trunc').stdout.split
  diff = all_images - instantiated_images

  describe diff do
    it { should be_empty }
  end
end

control 'cis-docker-benchmark-6.5' do
  impact 1.0
  title 'Avoid container sprawl'
  desc 'Do not keep a large number of containers on the same host.'

  tag 'host'
  tag cis: 'docker:6.5'
  tag level: 1
  ref 'https://zeltser.com/security-risks-and-benefits-of-docker-application/'
  ref 'http://searchsdn.techtarget.com/feature/Docker-networking-How-Linux-containers-will-change-your-network'

  total_on_host = command('docker info').stdout.split[1].to_i
  total_running = command('docker ps -q').stdout.split.length
  diff = total_on_host - total_running

  describe diff do
    it { should be <= MANAGEABLE_CONTAINER_NUMBER }
  end
end
